ITE/4/README.adoc

```diff
@@ -0,0 +1,462 @@
+= ITE-4: Generic URI Schemes for in-toto
+:source-highlighter: pygments
+:toc: preamble
+:toclevels: 2
+ifdef::env-github[]
+:tip-caption: :bulb:
+:note-caption: :information_source:
+:important-caption: :heavy_exclamation_mark:
+:caution-caption: :fire:
+:warning-caption: :warning:
+endif::[]
+
+.Metadata
+[cols="2"]
+|===
+| ITE
+| 4
+
+| Title
+| Generic URI Schemes for in-toto
+
+| Sponsor
+| link:https://github.com/santiagotorres[Santiago Torres-Arias]
+
+| Status
+| Draft :speech_balloon:
+
+| Type
+| Standards
+
+| Created
+| 2019-09-30
+
+|===
+
+[[abstract]]
+== Abstract
+
+This ITE proposes a generic Unique Resource Identifier (URI) scheme for in-toto
+artifacts. Currently, the
+link:https://github.com/in-toto/docs/blob/master/in-toto-spec.md[in-toto specification]
+only allows for files in the local filesystem to be the artifacts of link
+metadata. This ITE defines the scheme for a generic URI and provides a
+description of the expected behaviour of in-toto implementations that support
+generic URIs.
+
+[[specification]]
+== Specification
+
+The current in-toto specification must be modified to provide support for
+generic URI schemes. The current specification allows only for files in the
+local filesystem to be set as artifacts. With the addition of other entities,
+the parts of the specification that directly interact or handle the artifacts
+must be updated.
+
+in-toto link metadata files currently look like this:
+
+```
+{
+    "_type": "link",
+    "name": "<NAME>",
+    "command": "<COMMAND>",
+    "materials": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "products": {
+        "<PATH>": <HASH>,
+        "...": ...
+    },
+    "byproducts": {
+        "stderr": "<STDERR>",
+        "stdout": "<STDOUT>",
+        "return-value": "<RETURN-VALUE>"
+    },
+    "environment": {
+        "variables": "<ENV>",
+        "filesystem": "<FS>",
+        "workdir": "<PWD>"
+    }
+}
+```
+
+The `materials` and `products` fields contain key-value pairs. The URI for each
+artifact is its path in the filesystem. With this ITE active, the artifact could
```

Comment:
> artifact is its path in the filesystem. With this ITE active, the artifact could
 Is this "could/MAY" or "MUST"?  In other words, are you proposing to change /tmp/buildfile/a.out to be file:///tmp/buildfile/a.out

Reply:
I'll update the phrasing. The idea is for current filesystem paths to co-exist with generic URIs. The filesystem URIs _can_ have `file://` prepended, but it's not "MUST". The important bit here is that if it is a generic URI, it MUST follow the RFC's format.

Reply:
> Such a generic URI MUST have a fixed structure and MUST contain all the information necessary for the verifier to resolve if any part of the verification workflow requires it.

Does this at 81565d3 clarify?

Reply:
Minor nit: I would also reword/uppercase ...
```suggestion
artifact is its path in the filesystem. With this ITE active, the artifact CAN
```

I think then it gets even clearer. :)

Reply:
This isn't part of the RFC. Is it okay to use the same format?

Reply:
Good catch and apologies for the nonsense suggestion. Maybe *MAY* would be the right choice of words?